django/contrib/admin/widgets.py

```diff
@@ -242,7 +242,7 @@ def format_value(self, value):
         return ",".join(str(v) for v in value) if value else ""
 
 
-class RelatedFieldWidgetWrapper(forms.Widget):
+class RelatedFieldWidgetWrapper(forms.widgets.ChoiceWidget):
```

Comment:
Yes, no, and yes 😅

Will split it out and do an audit of the methods.

Reply:
Huh, so I was going to split this out into a separate PR, but I encountered some issues. It seems that this commit is a bit chicken and egg with the other one. The issue is that doing this first causes database queries to happen early which causes a bunch of tests to fail.

<details>
<summary>Test failures due to unexpected database access.</summary>

```
======================================================================
FAIL: test_ForeignKey (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_ForeignKey)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 128, in test_ForeignKey
    self.assertFormfield(Event, "main_band", forms.Select)
  File "/tests/django/tests/admin_widgets/tests.py", line 86, in assertFormfield
    ff = ma.formfield_for_dbfield(model._meta.get_field(fieldname), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_filtered_many_to_many (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_filtered_many_to_many)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 167, in test_filtered_many_to_many
    self.assertFormfield(
  File "/tests/django/tests/admin_widgets/tests.py", line 86, in assertFormfield
    ff = ma.formfield_for_dbfield(model._meta.get_field(fieldname), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_formfield_overrides_m2m_filter_widget (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_formfield_overrides_m2m_filter_widget)
The autocomplete_fields, raw_id_fields, filter_vertical, and
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 212, in test_formfield_overrides_m2m_filter_widget
    field = ma.formfield_for_dbfield(Band._meta.get_field("members"), request=None)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_m2m_widgets (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_m2m_widgets)
m2m fields help text as it applies to admin app (#9321).
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 265, in test_m2m_widgets
    self.assertFormfield(
  File "/tests/django/tests/admin_widgets/tests.py", line 86, in assertFormfield
    ff = ma.formfield_for_dbfield(model._meta.get_field(fieldname), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_m2m_widgets_no_allow_multiple_selected (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_m2m_widgets_no_allow_multiple_selected)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 288, in test_m2m_widgets_no_allow_multiple_selected
    self.assertFormfield(
  File "/tests/django/tests/admin_widgets/tests.py", line 86, in assertFormfield
    ff = ma.formfield_for_dbfield(model._meta.get_field(fieldname), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_many_to_many (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_many_to_many)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 159, in test_many_to_many
    self.assertFormfield(Band, "members", forms.SelectMultiple)
  File "/tests/django/tests/admin_widgets/tests.py", line 86, in assertFormfield
    ff = ma.formfield_for_dbfield(model._meta.get_field(fieldname), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_radio_fields_ForeignKey (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_radio_fields_ForeignKey)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 139, in test_radio_fields_ForeignKey
    ff = self.assertFormfield(
         ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/tests/admin_widgets/tests.py", line 86, in assertFormfield
    ff = ma.formfield_for_dbfield(model._meta.get_field(fieldname), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

======================================================================
FAIL: test_radio_fields_foreignkey_formfield_overrides_empty_label (admin_widgets.tests.AdminFormfieldForDBFieldTests.test_radio_fields_foreignkey_formfield_overrides_empty_label)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/admin_widgets/tests.py", line 155, in test_radio_fields_foreignkey_formfield_overrides_empty_label
    ff = ma.formfield_for_dbfield(Inventory._meta.get_field("parent"), request=None)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/options.py", line 205, in formfield_for_dbfield
    formfield.widget = widgets.RelatedFieldWidgetWrapper(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/contrib/admin/widgets.py", line 263, in __init__
    super().__init__(widget.attrs, widget.choices)
  File "/tests/django/django/forms/widgets.py", line 626, in __init__
    self.choices = list(choices)
                   ^^^^^^^^^^^^^
  File "/tests/django/django/forms/models.py", line 1418, in __len__
    return self.queryset.count() + (1 if self.field.empty_label is not None else 0)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/query.py", line 600, in count
    return self.query.get_count(using=self.db)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 570, in get_count
    return obj.get_aggregation(using, {"__count": Count("*")})["__count"]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/query.py", line 556, in get_aggregation
    result = compiler.execute_sql(SINGLE)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/db/models/sql/compiler.py", line 1560, in execute_sql
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tests/django/django/test/testcases.py", line 165, in __call__
    raise DatabaseOperationForbidden(self.message)
django.test.testcases.DatabaseOperationForbidden: Database queries to 'default' are not allowed in SimpleTestCase subclasses. Either subclass TestCase or TransactionTestCase to ensure proper test isolation or add 'default' to admin_widgets.tests.AdminFormfieldForDBFieldTests.databases to silence this failure.

----------------------------------------------------------------------
```

</details>

This seems to be then resolved by the main commit. And the main commit requires this commit to avoid a performance issue as a result of the way that `RelatedFieldWidgetWrapper` behaves currently.

I did also look at the stuff that is inherited and noted that I couldn't entirely remove `RelatedFieldWidgetWrapper.__deepcopy__()` as we still need to deepcopy the wrapped widget for which a reference is stored. Further to that, it seems that we cannot just eliminate other things, e.g. `is_hidden()` as then tests break because we're expecting things to be handed off to the wrapped widget.

Instead I've made `RelatedFieldWidgetWrapper.choices` hand off to the underlying widget and just gone back to re-normalizing on the widget anyway. _(You can see a few approaches I've tried which we can squash away.)_